policies/recipes/steps.go

```diff
@@ -90,8 +99,344 @@ func parsePermissions(s string) (os.FileMode, error) {
+			return err
+		}
+		return extractTar(decompressed.Name(), step.ArchiveExtraction.Destination)
+	default:
+		return fmt.Errorf("Unrecognized archive type %q", step.ArchiveExtraction.GetType())
+	}
+}
+
+func zipIsDir(name string) bool {
+	return strings.HasSuffix(name, "/")
+}
+
+func extractZip(zipPath string, dst string) error {
+	zr, err := zip.OpenReader(zipPath)
+	if err != nil {
+		return err
+	}
+	defer zr.Close()
+
+	// Check for conflicts
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.FileHeader.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if zipIsDir(f.Name) && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	// Create dirs
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.Name))
+		if err != nil {
+			return err
+		}
+
+		if !zipIsDir(f.Name) {
+			continue
+		}
+		_, err = os.Stat(filen)
+		if err == nil {
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return err
+		}
+		mode := f.Mode()
+		if mode == 0 {
+			mode = 0755
+		}
+		err = os.MkdirAll(filen, mode)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create files.
+	for _, f := range zr.File {
+		filen, err := common.NormPath(filepath.Join(dst, f.Name))
+		if err != nil {
+			return err
+		}
+		if zipIsDir(f.Name) {
+			continue
+		}
+		filedir := filepath.Dir(filen)
+		_, err = os.Stat(filedir)
+		if os.IsNotExist(err) {
+			err = os.MkdirAll(filedir, 0755)
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Printf("os.Create %s\n", filen)
+		reader, err := f.Open()
+		if err != nil {
+			return err
+		}
+
+		mode := f.Mode()
+		if mode == 0 {
+			mode = 0755
+		}
+
+		dst, err := os.OpenFile(filen, os.O_RDWR|os.O_CREATE, mode)
+		if err == nil {
+			_, err = io.Copy(dst, reader)
+		}
+		reader.Close()
+
+		if err != nil {
+			return err
+		}
+		err = os.Chtimes(filen, time.Now(), f.ModTime())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func extractTar(tarName string, dst string) error {
+	file, err := os.Open(tarName)
+	if err != nil {
+		return err
+	}
+	tr := tar.NewReader(file)
+
+	// Check for conflicts
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		filen, err := common.NormPath(filepath.Join(dst, header.Name))
+		if err != nil {
+			return err
+		}
+		fmt.Printf("checking if file %s exists\n", filen)
+		stat, err := os.Stat(filen)
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if header.Typeflag == tar.TypeDir && stat.IsDir() {
+			// it's ok if directories already exist
+			continue
+		}
+		return fmt.Errorf("file exists: %s", filen)
+	}
+
+	file.Seek(0, 0)
+	tr = tar.NewReader(file)
+
+	// Create dirs
+	for {
```

Comment:
You shouldn't need to loop through twice, just create directories and files in the same loop.

Reply:
Done